### ISSUE TITLE
samtools index can't operate on stdin

### ISSUE THREAD

**Issue by jrandall** [samtools/samtools#427]
It would be useful if `samtools index` could operate in a streaming manner. The easiest way to do this would probably be to allow `samtools index` to operate on stdin (input file '-') and to add an option to specify the output filename (or perhaps the prefix) for the index file. 

This could be used in a pipeline like:
`samtools sort unsorted.bam | tee sorted.bam | samtools index -p sorted.bam -`

The reliance on `tee` could perhaps be avoided by implementing a pass-through in `samtools index`:
`samtools sort unsorted.bam | samtools index -o sorted.bam -` (which would write both `sorted.bam` and `sorted.bam.bai`)

**Comment by jkbonfield** [comment]
It does I think already, but it needs the newer interface to allow the output filename to be generated (unsupported at the moment).

```
jkb@seq3a[samtools.../samtools] cat /nfs/users/nfs_j/jkb/scratch/data/SRR065390_1.srt.cram | ./samtools index -
jkb@seq3a[samtools.../samtools] ./samtools index /nfs/users/nfs_j/jkb/scratch/data/SRR065390_1.srt.cram
jkb@seq3a[samtools.../samtools] cmp /nfs/users/nfs_j/jkb/scratch/data/SRR065390_1.srt.cram.crai ./-.crai
```

```
jkb@seq3a[samtools.../samtools] cat /nfs/users/nfs_j/jkb/scratch/data/SRR065390_1.srt.bam | ./samtools index -
jkb@seq3a[samtools.../samtools] ./samtools index /nfs/users/nfs_j/jkb/scratch/data/SRR065390_1.srt.bam
jkb@seq3a[samtools.../samtools] cmp /nfs/users/nfs_j/jkb/scratch/data/SRR065390_1.srt.bam.bai ./-.bai
```

So it works.  However samtools *claims* it supports this in a better manner:

```
jkb@seq3a[samtools.../samtools] ./samtools index
Usage: samtools index [-bc] [-m INT] <in.bam> [out.index]
Options:
  -b       Generate BAI-format index for BAM files [default]
  -c       Generate CSI-format index for BAM files
  -m INT   Set minimum interval size for CSI indices to 2^INT [14]
jkb@seq3a[samtools.../samtools] cat /nfs/users/nfs_j/jkb/scratch/data/SRR065390_1.srt.cram | ./samtools index - /nfs/users/nfs_j/jkb/scratch/data/SRR065390_1.srt.cram.crai
Samtools-htslib-API: bam_index_build2() not yet implemented
Aborted (core dumped)
```

So basically it's just bam_index_build2() that we need finishing off.

Your last point is solid though - it ought to output both together.

**Comment by jrandall** [comment]
This workaround is what I'm using in place of this functionality (input is `input.bam`, output is `indexed.bam` and `indexed.bam.bai`): 
```
rm -f ${TMPDIR}/indexed.fifo.bam ${TMPDIR}/indexed.fifo.bam.bai; mkfifo ${TMPDIR}/indexed.fifo.bam; mkfifo ${TMPDIR}/indexed.fifo.bam.bai; (cat input.bam | tee indexed.bam ${TMPDIR}/indexed.fifo.bam > /dev/null)& (cat ${TMPDIR}/indexed.fifo.bam.bai > indexed.bam.bai)& samtools index ${TMPDIR}/indexed.fifo.bam
```

**Comment by jkbonfield** [comment]
I just tested this again and I don't see the issue.  Maybe it's been fixed since the issue was created.  Can you please test and if it works for you, close this?  Thanks.

```./samtools sort -o /tmp/a.bam - | tee srt.bam | ./samtools index - srt.bam.bai```

**Comment by jrandall** [comment]
Fixed by #8f60492d1a81b362cf7630ee3ab8669fd21420e0
